Implement System Configuration management and access control

geovanedcs opened the pull request:
### Description of the Change

This pull request refactors the access control for `SystemConfig` endpoints and includes email domain validation. Additionally, it implements the necessary CRUD operations for managing System Configuration.

### Benefits

- Improved access control security for `SystemConfig` endpoints.
- Enhanced validation for email domains to ensure compliance.
- Full support for creating, reading, updating, and deleting System Configuration.

### Possible Drawbacks

None identified.

### Verification Process

- Verified access control restrictions for all `SystemConfig` endpoints.
- Tested email domain validation with various valid and invalid emails.
- Assured CRUD operations work correctly through functional tests.

### Applicable Issues

None.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Endpoints /config para visualizar, salvar e remover a configuração do sistema (acesso restrito a administradores).
  * Validação de e‑mail que exige domínio @utfpr.edu.br; entradas inválidas retornam 400.
  * CORS configurado com origens definidas por ambiente.

* **Tests**
  * Novos testes de integração e unitários para controladora, serviço e validador de e‑mail.

* **Chores**
  * Migração cria tabela de configuração com índice garantindo única configuração ativa e insere linha padrão.
  * Dependências adicionadas para validação Jakarta e testes de segurança.
* **Docs**
  * JavaDoc adicionada à configuração de segurança e CORS.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

geovanedcs committed:
Implement System Configuration management with CRUD operations

geovanedcs committed:
Refactor access control for SystemConfig endpoints and validate email domain

geovanedcs committed:
Add deleteConfig method and enhance email validation in SystemConfig

geovanedcs committed:
Add is_active field to system_config table and update related SQL insert logic

geovanedcs committed:
Update saveConfig method to initialize existingConfig with default ID

geovanedcs committed:
Refactor saveConfig method and update system_config table to use identity for ID

geovanedcs committed:
Add is_active field to SystemConfig and implement retrieval of active configuration

geovanedcs committed:
Merge branch 'dev' of https://github.com/utfprpb-oficina202502/laboratorio-dainf-server into feature/SCRUM-39-Criar-configuracoes

# Conflicts:
#	src/main/java/br/com/utfpr/gerenciamento/server/dto/dashboards/DashboardEmprestimoCountRangeResponseDto.java
#	src/main/java/br/com/utfpr/gerenciamento/server/security/WebSecurity.java
#	src/main/java/br/com/utfpr/gerenciamento/server/service/impl/DashboardServiceImpl.java
#	src/main/java/br/com/utfpr/gerenciamento/server/service/impl/EmprestimoServiceImpl.java
#	src/main/java/br/com/utfpr/gerenciamento/server/service/impl/UsuarioServiceImpl.java
#	src/main/resources/db/common/V2.3__Create_System_Config_Table.sql
#	src/main/resources/db/test/V2.3__Create_System_Config_Table.sql

geovanedcs committed:
feat - add system_config table with is_active field and update security configuration

geovanedcs commented on src/test/java/br/com/utfpr/gerenciamento/server/controller/SystemConfigControllerTest.java:
```diff
@@ -0,0 +1,110 @@
+class SystemConfigControllerTest {
+  @Autowired private WebApplicationContext context;
+
+  @Spy private SystemConfigService service;
```
Foi utilizado o spy pois o MockBean foi deprecated no spring 3.4

geovanedcs replied:
@coderabbitai Foi utilizado o spy pois o MockBean foi deprecated no spring 3.4

geovanedcs committed:
refactor - update SystemConfigControllerTest to use MockitoBean and simplify security configuration

geovanedcs committed:
refactor - simplify authentication in SystemConfigControllerTest using SecurityMockMvcRequestPostProcessors

geovanedcs committed:
refactor - simplify formatting and improve readability in SystemConfigControllerTest

rodizipa committed:
devops - perfil de testes

rodizipa committed:
bugfix - adicionado perfil de teste para teste que falha

geovanedcs committed:
test - add unit tests for SystemConfigServiceImpl and UtfprEmailValidator

geovanedcs commented:
@coderabbitai generate docstrings

rodizipa commented on src/main/java/br/com/utfpr/gerenciamento/server/controller/SystemConfigController.java:
```diff
@@ -0,0 +1,39 @@
+  @PreAuthorize("hasAuthority('ROLE_ADMINISTRADOR')")
+  public ResponseEntity<SystemConfig> saveConfig(@Valid @RequestBody SystemConfig config) {
+    String email = config.getNadaConstaEmail();
+    if (email == null || !email.endsWith("@utfpr.edu.br")) {
```
Essa validação aqui não faz sentido se já tem um validador, coloca nas entidades e tira isso daqui, controller não é o lugar pra isso.

coderabbitai[bot] committed:
📝 Add docstrings to `feature/SCRUM-39-Criar-configuracoes`

Docstrings generation was requested by @geovanedcs.

* https://github.com/utfprpb-oficina202502/laboratorio-dainf-server/pull/9#issuecomment-3406496410

The following files were modified:

* `src/main/java/br/com/utfpr/gerenciamento/server/controller/SystemConfigController.java`
* `src/main/java/br/com/utfpr/gerenciamento/server/repository/SystemConfigRepository.java`
* `src/main/java/br/com/utfpr/gerenciamento/server/security/WebSecurity.java`
* `src/main/java/br/com/utfpr/gerenciamento/server/service/SystemConfigService.java`
* `src/main/java/br/com/utfpr/gerenciamento/server/service/impl/SystemConfigServiceImpl.java`
* `src/main/java/br/com/utfpr/gerenciamento/server/util/UtfprEmailValidator.java`

rodizipa commented on src/main/java/br/com/utfpr/gerenciamento/server/util/UtfprEmailValidator.java:
```diff
@@ -0,0 +1,40 @@
+    @Override
+    public boolean isValid(String value, ConstraintValidatorContext context) {
+      if (value == null) return false;
+      String trimmed = value.trim();
```
normaliza o email também, deixa tudo lowercase

geovanedcs committed:
Merge pull request #11 from utfprpb-oficina202502/coderabbitai/docstrings/25ed356

📝 Add docstrings to `feature/SCRUM-39-Criar-configuracoes`

geovanedcs committed:
feat - add config endpoints with restricted access for administrators

rodizipa commented on src/main/java/br/com/utfpr/gerenciamento/server/util/UtfprEmailValidator.java:
```diff
@@ -0,0 +1,40 @@
+      int atIdx = trimmed.lastIndexOf('@');
+      if (atIdx < 0 || atIdx == trimmed.length() - 1) return false;
+      String domain = trimmed.substring(atIdx + 1).toLowerCase();
+      return domain.equals("utfpr.edu.br");
```
- Bota essa string em uma constante, evita texto mágico no código
- O código n valida se tem quebra de linha, se eu por um \n\r já dá ruim

rodizipa commented on src/main/java/br/com/utfpr/gerenciamento/server/service/impl/SystemConfigServiceImpl.java:
```diff
@@ -0,0 +1,44 @@
+            .orElseGet(
+                () -> {
+                  SystemConfig sc = new SystemConfig();
+                  sc.setId(1L);
```
Por qual razão está sendo configurado esse ID fixo?

rodizipa reviewed (changes requested):
Ajustar os comentários

geovanedcs committed:
refactor - improve docstring formatting and readability across multiple classes

geovanedcs committed:
feat - implement UtfprEmailValidator for email validation in SystemConfig

geovanedcs committed:
feat - implement UtfprEmailValidator for email validation in SystemConfig

geovanedcs committed:
test - enhance SystemConfigServiceImplTest with ArgumentCaptor for saveConfig verification

rodizipa reviewed (approved):
Revisar os temas do coderabbit

geovanedcs committed:
test - remove unnecessary Spring annotations from UtfprEmailValidatorTest
